Add changelog entries for 11.1.0 (#6349)
citusdata/citus: CHANGELOG.md

### citus v11.1.0 (September 15, 2022) ###

* Adds support for PostgreSQL 15beta3

* Adds a rebalancer that uses background tasks for its execution

* Adds `create_distributed_table_concurrently()` which distributes tables
  without blocking writes

> * Adds `citus_split_shard_by_split_points()` function

Reviewer: Also a missing item / 2 items: 
```
* `isolate_tenant_to_new_shard` no longer blocks writes
* Adds a shard_transfer_mode argument to `isolate_tenant_to_new_shard` 
```

> * Adds `citus_split_shard_by_split_points()` function

Author: Details written for create_distributed_table_concurrently() seemed enough to me, so I would like to hear from you how I can improve the entry related to it further.

On the other hand for citus_split_shard_by_split_points(), do you think the following sounds good ?:
> Adds `citus_split_shard_by_split_points()` function that allows splitting a shard to specified set of nodes without blocking writes and based on given split points

> * Adds `create_distributed_table_concurrently()` which distributes tables
>   without blocking writes
> 
> * Adds `citus_split_shard_by_split_points()` function

Reviewer 2: > Details written for create_distributed_table_concurrently() seemed enough to me.

On second thought, I agree with you. Let's leave it as it is.

This suggestion is quite nice overall. I changed function into UDF and come up with the following:

```suggestion
* Adds `create_distributed_table_concurrently()` which distributes tables
  without blocking writes

* Adds `citus_split_shard_by_split_points()` UDF that allows
  splitting a shard to specified set of nodes without blocking writes
  and based on given split points
```

* Adds support for non-blocking tenant isolation

* Separates columnar storage AM into a separate logical extension

> * Supports logical replication in `replicate_reference_tables()`

Reviewer 2: ```suggestion
* Adds support for logical replication in `replicate_reference_tables()`
```

* Improves performance of blocking shard moves

> * Uses a faster custom copy logic for non-blocking shard moves

Reviewer 2: ```suggestion
* Improves non-blocking shard moves with a faster custom copy logic
```

We do not have much `Uses ..` entries. This is an improvement to an already existing feature I guess.

Author: agreed

* `isolate_tenant_to_new_shard()` adds support for columnar tables

* `isolate_tenant_to_new_shard()` adds support for partitioned tables

* `isolate_tenant_to_new_shard()` drops support for replicated tables

* `isolate_tenant_to_new_shard()` now fails when run concurrently with itself

> 

Reviewer 2: We have about 6 entries for tenant isolation. Can we group them together?

Reviewer 2: Also, these 4 entries do not start with a verb.

```suggestion
* Adds support for isolating tenants that use columnar tables

* Adds support for isolating tenants that use partitioned tables

* Drops support for isolating tenants that use replicated tables

* Prevents concurrent tenant isolation operations

```

Reviewer 2: I realized that we already have an entry like the following on Enterprise 11.0.0 beta

> Prevents concurrent tenant isolations

https://github.com/citusdata/citus-enterprise/blob/enterprise-master/CHANGELOG.md#citus-enterprise-v1100_beta-march-22-2022

I believe this is not actually something new on Citus 11.1. Am I missing something here?

Author: > I realized that we already have an entry like the following on Enterprise 11.0.0 beta

You're right, I should drop this one, but not sure about the others. But it's true that others at least should be grouped.

Author: but will apply your suggestions for other three

> * Adds an `allow_unsafe_constraints` flag for constraints without distribution
>   column

Reviewer 2: `allow_unsafe_constraints` is not a flag but a GUC.

```suggestion
* Adds `allow_unsafe_constraints` GUC for constraints without distribution
  column
```

> * Adds support for `GRANT/REVOKE` ON aggregates

Author: ```suggestion
* Adds support for `GRANT/REVOKE` on aggregates
```

* Adds support for `NULLS NOT DISTINCT` clauses for indexes

* Adds support for setting relation options for columnar tables using
  `ALTER TABLE`

* Adds support for unlogged distributed sequences

* Checks existence of the shards before insert, delete, and update

* Creates all foreign keys quickly at the end of a shard move

* Hides tables owned by extensions from `citus_tables` and `citus_shards`

* Introduces GUC `citus.skip_constraint_validation`

* Introduces `citus_locks` view

* Propagates `VACUUM` and `ANALYZE` to worker nodes

* Makes non-partitioned table size calculation quicker

> * Only shows shards in moving state in `get_rebalance_progress()`

Reviewer 2: Does not start with a verb. We can find a better alternative than my suggestion below.

```suggestion
* Limits `get_rebalance_progress()` to show shards in moving state
```

* Removes `do_repair` option from `citus_copy_shard_placement()`

* Removes deprecated re-partitioning functions like
  `worker_hash_partition_table()`

* Replaces `citus.hide_shards_from_app_name_prefixes` GUC with
  `citus.show_shards_for_app_name_prefixes`

* Shows `citus_copy_shard_placement()` progress in `get_rebalance_progres()`

> * Shows local tables added to metadata in `citus_tables`

Reviewer 2: ```suggestion
* Improves `citus_tables` by showing local tables added to metadata
```

* Supports changing CPU priorities for backends and shard moves

> * Tightens security and improves visibility for columnar table access method

Reviewer 2: Personally I do not understand what it means to improve visibility for columnar table AM. Without understanding what it means, I am unable to propose a better alternative.

Author: I've read about the PR now and here is my proposal:

```
* Improves columnar table access method by moving old catalog tables into
  an internal schema and introduces more secure & informative views based
  on them
```

> * `citus_move_shard_placement()` becomes a noop if shard already exists on node

Reviewer 2: Does not start with a verb. It is tough to rewrite this one.

> * `create_distributed_table()` creates new colocation entries when using
>   `colocate_with => 'none'`

Reviewer 2: ```suggestion
* Improves `create_distributed_table()` by creating new colocation entries when
  using `colocate_with => 'none'`
```

> * Makes sure that `SELECT . FOR UPDATE `opens a transaction block when used in

Reviewer 2: ```suggestion
* Makes sure that `SELECT .. FOR UPDATE `opens a transaction block when used in
```

Author: There are some other occurences of `X .. Y` in the changelog, so I think leaving this as-is should be okay.

Reviewer 2: Please ignore this comment, as I wrote a better alternative in a second round of reviews here: https://github.com/citusdata/citus/pull/6349#discussion_r971891341

> * Makes sure that `SELECT . FOR UPDATE `opens a transaction block when used in
>   a function call

Reviewer 2: ```suggestion
* Ensures that `SELECT .. FOR UPDATE` opens a transaction block when used in
  a function call
```

> * Makes sure to disallow usage of SQL functions referencing to a distributed
>   table and prevents a segfault

Reviewer 2: ```suggestion
* Prevents a segfault by disallowing usage of SQL functions referencing to a
  distributed table
```

* Prevents creating a new colocation entry when replicating reference tables

* Fixes a bug in query escaping in `undistribute_table()` and
  `alter_distributed_table()`

* Fixes a bug preventing the usage of `isolate_tenant_to_new_shard()` with text
  column

> * Fixes a bug that caused `GRANT` to propagate within `CREATE EXTENSION`

Reviewer 2: ```suggestion
* Fixes a bug that may cause `GRANT` to propagate within `CREATE EXTENSION`
```

* Fixes a bug that causes incorrectly marking `metadatasynced` flag for
  coordinator

> * Fixes a bug that may cause Citus not to create function in transaction block
>   properly

Reviewer 2: ```suggestion
* Fixes a bug that may prevent Citus from creating function in transaction block
  properly
```

IMO, the original was harder to comprehend.

* Fixes a bug that prevents promoting read-replicas as primaries

* Fixes a bug that prevents setting colocation group of a partitioned
  distributed table to `none`

> * Fixes a bug that prevents using `auto` option for `VACUUM (INDEX_CLEANUP)`

Reviewer 2: ```suggestion
* Fixes a bug that prevents using `AUTO` option for `VACUUM (INDEX_CLEANUP)`
```

This is part of the SQL syntax. I guess it is better to have it in caps as the PG docs does it.

### citus v11.1.0 (September 15, 2022) ###

* Adds support for PostgreSQL 15beta4

> * Adds a rebalancer that uses background tasks for its execution

Reviewer 3: Hmm, `Adds` sounds a little odd. 

Maybe: `Adds ability to run shard rebalancer in the background` or such?

> * Adds `create_distributed_table_concurrently()` UDF to distribute tables

Reviewer 3: I think Marco's blog refers this as `without interrupting the application` instead of `without blocking writes`

  without blocking writes

* Adds `citus_split_shard_by_split_points()` UDF that allows
  splitting a shard to specified set of nodes without blocking writes
  and based on given split points

* Adds support for non-blocking tenant isolation

* Adds support for isolation tenants that use partitioned tables
  or columnar tables

* Drops support for isolation tenants that use replicated tables

* Separates columnar table access method into a separate logical extension

> * Adds support for logical replication in `replicate_reference_tables()`

Reviewer 3: This sounds too technical, maybe: `Adds support for online replication in `replicate_reference_tables()``

* Improves performance of blocking shard moves

* Improves non-blocking shard moves with a faster custom copy logic

> * Adds the GUC `allow_unsafe_constraints` to allow constraints without

Reviewer 3: Hmm, what does `constraints without distribution column` mean? Can you share the link to the PR that adds this?

Reviewer 3: also, maybe `citus.allow_unsafe_constraints`

Author: https://github.com/citusdata/citus/pull/6237

Reviewer 3: oh, I haven't seen this. Maybe we call it as `unique constraint`? Because, for example, we don't seem to relax foreign keys etc

  distribution column

* Introduces GUC `citus.skip_constraint_validation`

> * Replaces `citus.hide_shards_from_app_name_prefixes` GUC with

Reviewer 3: Isn't this 11.0 item? https://docs.citusdata.com/en/v11.0/develop/api_guc.html?highlight=show_shards_for_app_name_prefixes#citus-show-shards-for-app-name-prefixes-text

 And, that was an internal change within the release, so I don't think we should have this here

  `citus.show_shards_for_app_name_prefixes`

* Introduces `citus_locks` view

* Improves `citus_tables` view by showing local tables added to metadata

* Improves columnar table access method by moving old catalog tables into
  an internal schema and introduces more secure & informative views based
  on them

> * Removes `do_repair` option from `citus_copy_shard_placement()`

Reviewer 3: As similar to earlier change log entries, can't we group `Adds` and `Removes` back to back, instead of scattered around?

For example, `Drops support for isolation tenants that use replicated tables` should also be close to here (and if we have ay others).

And, also all `removes/drops` could come after all `adds` and before `fixes`

* Removes deprecated re-partitioning functions like
  `worker_hash_partition_table()`

* Adds support for `GRANT/REVOKE` on aggregates

> * Adds support for `NULLS NOT DISTINCT` clauses for indexes

Reviewer 3: maybe add `for Postgres 15+` or such

* Adds support for setting relation options for columnar tables using
  `ALTER TABLE`

* Adds support for unlogged distributed sequences

* Checks existence of the shards before insert, delete, and update

* Creates all foreign keys quickly at the end of a shard move

* Hides tables owned by extensions from `citus_tables` and `citus_shards`

* Shows `citus_copy_shard_placement()` progress in `get_rebalance_progres()`

* Propagates `VACUUM` and `ANALYZE` to worker nodes

* Makes non-partitioned table size calculation quicker

* Limits `get_rebalance_progress()` to show shards in moving state

* Supports changing CPU priorities for backends and shard moves

> * Makes `citus_move_shard_placement()` idempotent if shard already exists

Reviewer 3: can we group all rebalancer related changes back to back

Potentially around the top of the change log

* Adds support for PostgreSQL 15beta3

* Adds a rebalancer that uses background tasks for its execution

* Adds `create_distributed_table_concurrently()` which distributes tables
  without blocking writes

* Adds `citus_split_shard_by_split_points()` function

* Adds support for non-blocking tenant isolation

* Separates columnar storage AM into a separate logical extension

* Supports logical replication in `replicate_reference_tables()`

* Improves performance of blocking shard moves

* Uses a faster custom copy logic for non-blocking shard moves

* `isolate_tenant_to_new_shard()` adds support for columnar tables

* `isolate_tenant_to_new_shard()` adds support for partitioned tables

* `isolate_tenant_to_new_shard()` drops support for replicated tables

* `isolate_tenant_to_new_shard()` now fails when run concurrently with itself

* Adds an `allow_unsafe_constraints` flag for constraints without distribution
  column

* Adds support for `GRANT/REVOKE` ON aggregates

* Adds support for `NULLS NOT DISTINCT` clauses for indexes

* Adds support for setting relation options for columnar tables using
  `ALTER TABLE`

* Adds support for unlogged distributed sequences

> * Adds the GUC `enable_unsupported_feature_messages` to control some of the
>   Citus related messages
> 

Author: ```suggestion
```
